Build for Python 3.11

Python 3.11 will be released towards the end of this year ([possibly October](https://peps.python.org/pep-0664/#schedule)). 

There is a release candidate available now (Python 3.11.0rc1, released 2022-08-29), which is expected to be a stable ABI with the final 3.11.0 release. Therefore, it is reasonable to prepare JPype against this now, and potentially make a release containing a Python 3.11 wheel already.

The changelog is available at https://docs.python.org/3.11/whatsnew/3.11.html#id6.

In terms of complexity, it seems that there are a few issues which will likely come up:

 * The [PyFrameObject](https://docs.python.org/3.11/c-api/frame.html#c.PyFrameObject) structure members have been removed from the public C API. This is [used in the exception handling](https://github.com/jpype-project/jpype/blob/82a7658277b36c628ea305ea705cd012d81471dd/native/common/jp_exception.cpp#L524) in JPype
 * [_PyObject_GC_Malloc](https://github.com/jpype-project/jpype/blob/82a7658277b36c628ea305ea705cd012d81471dd/native/python/pyjp_value.cpp#L46) has been removed in https://github.com/python/cpython/commit/8319114feedd2a5b77378bba24eb9fb2689c5033 - since these weren't public API, there is no obvious suggestion for what to use instead.


cc @Thrameos

> Not sure how we can support exception reporting without a replacement API for PyFrameObject.

Is this something we need to raise to the CPython mailing list perhaps? It would be quite uncomfortable to lose that altogether.

FYI, I did a quick build against Python 3.11 with:

(dockerfile)

```
FROM quay.io/pypa/manylinux2014_x86_64
RUN yum install -y \
    java-devel \
    ccache \
;
COPY . /opt/src
RUN cd /opt/src \
 && CC="ccache gcc" python3.11 setup.py build_ext --inplace \
;
```

```
docker build .
Step 1/4 : FROM quay.io/pypa/manylinux2014_x86_64
 ---> 0b5b5fbafd04
Step 2/4 : RUN yum install -y     java-devel     ccache ;
 ---> Using cache
 ---> ff7a6bf33d03
Step 3/4 : COPY . /opt/src
 ---> f5eea5b22b9f
Step 4/4 : RUN cd /opt/src  && CC="ccache gcc" python3.11 setup.py build_ext --inplace ;
 ---> Running in 1a67a85058db
/opt/_internal/cpython-3.11.0rc1/lib/python3.11/site-packages/setuptools/dist.py:530: UserWarning: Normalizing '1.4.1_dev0' to '1.4.1.dev0'
  warnings.warn(tmpl.format(**locals()))
running build_ext
Call build extensions
Jar cache is missing, using --enable-build-jar to recreate it.
  javac -cp lib/asm-8.0.1.jar -d build/temp.linux-x86_64-cpython-311/org.jpype/classes -g:none -source 1.8 -target 1.8 -encoding UTF-8 native/java/org/jpype/JPypeContext.java native/java/org/jpype/JPypeKeywords.java native/java/org/jpype/JPypeSignal.java native/java/org/jpype/JPypeUtilities.java native/java/org/jpype/PyExceptionProxy.java native/java/org/jpype/classloader/DynamicClassLoader.java native/java/org/jpype/classloader/JPypeClassLoader.java native/java/org/jpype/html/AttrGrammar.java native/java/org/jpype/html/AttrParser.java native/java/org/jpype/html/Html.java native/java/org/jpype/html/HtmlGrammar.java native/java/org/jpype/html/HtmlHandler.java native/java/org/jpype/html/HtmlParser.java native/java/org/jpype/html/HtmlTreeHandler.java native/java/org/jpype/html/HtmlWriter.java native/java/org/jpype/html/Parser.java native/java/org/jpype/javadoc/DomUtilities.java native/java/org/jpype/javadoc/Javadoc.java native/java/org/jpype/javadoc/JavadocException.java native/java/org/jpype/javadoc/JavadocExtractor.java native/java/org/jpype/javadoc/JavadocRenderer.java native/java/org/jpype/javadoc/JavadocTransformer.java native/java/org/jpype/manager/ClassDescriptor.java native/java/org/jpype/manager/MethodResolution.java native/java/org/jpype/manager/ModifierCode.java native/java/org/jpype/manager/TypeAudit.java native/java/org/jpype/manager/TypeFactory.java native/java/org/jpype/manager/TypeFactoryNative.java native/java/org/jpype/manager/TypeManager.java native/java/org/jpype/pickle/ByteBufferInputStream.java native/java/org/jpype/pickle/Decoder.java native/java/org/jpype/pickle/Encoder.java native/java/org/jpype/pkg/JPypePackage.java native/java/org/jpype/pkg/JPypePackageManager.java native/java/org/jpype/proxy/JPypeProxy.java native/java/org/jpype/ref/JPypeReference.java native/java/org/jpype/ref/JPypeReferenceNative.java native/java/org/jpype/ref/JPypeReferenceQueue.java native/java/org/jpype/ref/JPypeReferenceSet.java
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Copy file native/java/org/jpype/html/entities.txt build/temp.linux-x86_64-cpython-311/org.jpype/classes/org/jpype/html/entities.txt
  jar cvf build/lib.linux-x86_64-cpython-311/org.jpype.jar -C build/temp.linux-x86_64-cpython-311/org.jpype/classes .
added manifest
adding: org/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/JPypeContext.class(in = 9682) (out= 4714)(deflated 51%)
adding: org/jpype/PyExceptionProxy.class(in = 223) (out= 189)(deflated 15%)
adding: org/jpype/classloader/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/classloader/JPypeClassLoader.class(in = 2649) (out= 1413)(deflated 46%)
adding: org/jpype/classloader/DynamicClassLoader$1.class(in = 1876) (out= 790)(deflated 57%)
adding: org/jpype/classloader/DynamicClassLoader.class(in = 6303) (out= 3073)(deflated 51%)
adding: org/jpype/ref/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/ref/JPypeReferenceQueue.class(in = 3058) (out= 1433)(deflated 53%)
adding: org/jpype/ref/JPypeReferenceSet.class(in = 1628) (out= 927)(deflated 43%)
adding: org/jpype/ref/JPypeReferenceQueue$1.class(in = 185) (out= 143)(deflated 22%)
adding: org/jpype/ref/JPypeReferenceNative.class(in = 267) (out= 204)(deflated 23%)
adding: org/jpype/ref/JPypeReferenceQueue$Worker.class(in = 1560) (out= 748)(deflated 52%)
adding: org/jpype/ref/JPypeReference.class(in = 541) (out= 355)(deflated 34%)
adding: org/jpype/ref/JPypeReferenceSet$Pool.class(in = 690) (out= 405)(deflated 41%)
adding: org/jpype/JPypeContext$1.class(in = 480) (out= 311)(deflated 35%)
adding: org/jpype/JPypeSignal$1.class(in = 576) (out= 360)(deflated 37%)
adding: org/jpype/JPypeKeywords.class(in = 1448) (out= 775)(deflated 46%)
adding: org/jpype/JPypeSignal.class(in = 1711) (out= 820)(deflated 52%)
adding: org/jpype/pkg/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/pkg/JPypePackage.class(in = 4038) (out= 2090)(deflated 48%)
adding: org/jpype/pkg/JPypePackageManager$ModuleDirectory.class(in = 1752) (out= 877)(deflated 49%)
adding: org/jpype/pkg/JPypePackageManager.class(in = 7815) (out= 3609)(deflated 53%)
adding: org/jpype/manager/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/manager/TypeManager$1.class(in = 177) (out= 135)(deflated 23%)
adding: org/jpype/manager/MethodResolution.class(in = 4502) (out= 2186)(deflated 51%)
adding: org/jpype/manager/ClassDescriptor.class(in = 1192) (out= 708)(deflated 40%)
adding: org/jpype/manager/TypeManager$Destroyer.class(in = 1116) (out= 626)(deflated 43%)
adding: org/jpype/manager/TypeManager.class(in = 15280) (out= 7263)(deflated 52%)
adding: org/jpype/manager/TypeFactory.class(in = 644) (out= 332)(deflated 48%)
adding: org/jpype/manager/TypeFactoryNative.class(in = 772) (out= 414)(deflated 46%)
adding: org/jpype/manager/ModifierCode.class(in = 2632) (out= 1425)(deflated 45%)
adding: org/jpype/manager/TypeAudit.class(in = 251) (out= 172)(deflated 31%)
adding: org/jpype/pickle/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/pickle/Decoder.class(in = 683) (out= 406)(deflated 40%)
adding: org/jpype/pickle/Encoder.class(in = 710) (out= 421)(deflated 40%)
adding: org/jpype/pickle/ByteBufferInputStream.class(in = 1179) (out= 649)(deflated 44%)
adding: org/jpype/JPypeUtilities.class(in = 779) (out= 444)(deflated 43%)
adding: org/jpype/proxy/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/proxy/JPypeProxy.class(in = 2535) (out= 1254)(deflated 50%)
adding: org/jpype/html/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/html/AttrGrammar$AttrRule.class(in = 1223) (out= 641)(deflated 47%)
adding: org/jpype/html/AttrGrammar.class(in = 2221) (out= 984)(deflated 55%)
adding: org/jpype/html/Html.class(in = 5891) (out= 3223)(deflated 45%)
adding: org/jpype/html/AttrGrammar$EndQuoteRule.class(in = 1130) (out= 564)(deflated 50%)
adding: org/jpype/html/HtmlGrammar$Comment.class(in = 1744) (out= 805)(deflated 53%)
adding: org/jpype/html/AttrGrammar$QuoteRule.class(in = 1087) (out= 545)(deflated 49%)
adding: org/jpype/html/HtmlGrammar$Quoted.class(in = 1638) (out= 748)(deflated 54%)
adding: org/jpype/html/HtmlGrammar$EndCData.class(in = 1648) (out= 834)(deflated 49%)
adding: org/jpype/html/entities.txt(in = 3616) (out= 1559)(deflated 56%)
adding: org/jpype/html/AttrGrammar$BooleanRule.class(in = 1209) (out= 634)(deflated 47%)
adding: org/jpype/html/Parser$MatchRule.class(in = 1073) (out= 605)(deflated 43%)
adding: org/jpype/html/HtmlGrammar$MergeText.class(in = 996) (out= 525)(deflated 47%)
adding: org/jpype/html/HtmlGrammar$StartElement.class(in = 1624) (out= 784)(deflated 51%)
adding: org/jpype/html/HtmlGrammar$Escaped.class(in = 1048) (out= 556)(deflated 46%)
adding: org/jpype/html/Parser$Token.class(in = 215) (out= 167)(deflated 22%)
adding: org/jpype/html/Parser$Rule.class(in = 351) (out= 197)(deflated 43%)
adding: org/jpype/html/HtmlGrammar$Token.class(in = 2243) (out= 1174)(deflated 47%)
adding: org/jpype/html/HtmlGrammar$Directive.class(in = 910) (out= 476)(deflated 47%)
adding: org/jpype/html/HtmlGrammar$StartQuote.class(in = 1259) (out= 611)(deflated 51%)
adding: org/jpype/html/AttrParser.class(in = 649) (out= 360)(deflated 44%)
adding: org/jpype/html/HtmlParser.class(in = 531) (out= 306)(deflated 42%)
adding: org/jpype/html/HtmlHandler.class(in = 352) (out= 218)(deflated 38%)
adding: org/jpype/html/HtmlTreeHandler.class(in = 5222) (out= 2595)(deflated 50%)
adding: org/jpype/html/AttrGrammar$Token.class(in = 1791) (out= 931)(deflated 48%)
adding: org/jpype/html/HtmlGrammar$BeginElement.class(in = 2130) (out= 962)(deflated 54%)
adding: org/jpype/html/Parser$Grammar.class(in = 254) (out= 165)(deflated 35%)
adding: org/jpype/html/AttrGrammar$IgnoreWSRule.class(in = 819) (out= 458)(deflated 44%)
adding: org/jpype/html/HtmlGrammar$State.class(in = 2080) (out= 905)(deflated 56%)
adding: org/jpype/html/Parser.class(in = 3984) (out= 1900)(deflated 52%)
adding: org/jpype/html/Parser$State.class(in = 349) (out= 198)(deflated 43%)
adding: org/jpype/html/HtmlGrammar$1.class(in = 171) (out= 136)(deflated 20%)
adding: org/jpype/html/HtmlGrammar.class(in = 5251) (out= 2249)(deflated 57%)
adding: org/jpype/html/HtmlGrammar$EndComment.class(in = 1670) (out= 853)(deflated 48%)
adding: org/jpype/html/HtmlGrammar$EndElement.class(in = 1392) (out= 664)(deflated 52%)
adding: org/jpype/html/HtmlGrammar$CompleteElement.class(in = 1709) (out= 853)(deflated 50%)
adding: org/jpype/html/Parser$Entity.class(in = 774) (out= 391)(deflated 49%)
adding: org/jpype/html/HtmlWriter.class(in = 3444) (out= 1779)(deflated 48%)
adding: org/jpype/html/HtmlGrammar$StartComment.class(in = 1554) (out= 723)(deflated 53%)
adding: org/jpype/html/Parser$1.class(in = 161) (out= 132)(deflated 18%)
adding: org/jpype/html/HtmlGrammar$Cleanup.class(in = 2482) (out= 1179)(deflated 52%)
adding: org/jpype/html/HtmlGrammar$EndDirective.class(in = 1404) (out= 677)(deflated 51%)
adding: org/jpype/html/HtmlGrammar$CData.class(in = 1700) (out= 777)(deflated 54%)
adding: org/jpype/html/AttrGrammar$State.class(in = 1830) (out= 776)(deflated 57%)
adding: org/jpype/javadoc/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/javadoc/JavadocTransformer.class(in = 9712) (out= 4497)(deflated 53%)
adding: org/jpype/javadoc/JavadocExtractor.class(in = 6448) (out= 3058)(deflated 52%)
adding: org/jpype/javadoc/JavadocException.class(in = 277) (out= 198)(deflated 28%)
adding: org/jpype/javadoc/Javadoc.class(in = 626) (out= 346)(deflated 44%)
adding: org/jpype/javadoc/DomUtilities.class(in = 3165) (out= 1450)(deflated 54%)
adding: org/jpype/javadoc/JavadocTransformer$Workspace.class(in = 916) (out= 430)(deflated 53%)
adding: org/jpype/javadoc/JavadocRenderer.class(in = 7226) (out= 3569)(deflated 50%)
Call build ext
copying build/lib.linux-x86_64-cpython-311/org.jpype.jar -> 
copying build/lib.linux-x86_64-cpython-311/_jpype.cpython-311-x86_64-linux-gnu.so -> 
Removing intermediate container 1a67a85058db
 ---> d58cc94938dc
Successfully built d58cc94938dc
(base) pelson@pelson-X1:~/github/jpype/jpype$ docker build .
Sending build context to Docker daemon  129.9MB
Step 1/4 : FROM quay.io/pypa/manylinux2014_x86_64
 ---> 0b5b5fbafd04
Step 2/4 : RUN yum install -y     java-devel     ccache ;
 ---> Using cache
 ---> ff7a6bf33d03
Step 3/4 : COPY . /opt/src
 ---> 4704e5325ca3
Step 4/4 : RUN cd /opt/src  && CC="ccache gcc" python3.11 setup.py build_ext --inplace ;
 ---> Running in 714e9067964b
/opt/_internal/cpython-3.11.0rc1/lib/python3.11/site-packages/setuptools/dist.py:530: UserWarning: Normalizing '1.4.1_dev0' to '1.4.1.dev0'
  warnings.warn(tmpl.format(**locals()))
running build_ext
Call build extensions
Jar cache is missing, using --enable-build-jar to recreate it.
  javac -cp lib/asm-8.0.1.jar -d build/temp.linux-x86_64-cpython-311/org.jpype/classes -g:none -source 1.8 -target 1.8 -encoding UTF-8 native/java/org/jpype/JPypeContext.java native/java/org/jpype/JPypeKeywords.java native/java/org/jpype/JPypeSignal.java native/java/org/jpype/JPypeUtilities.java native/java/org/jpype/PyExceptionProxy.java native/java/org/jpype/classloader/DynamicClassLoader.java native/java/org/jpype/classloader/JPypeClassLoader.java native/java/org/jpype/html/AttrGrammar.java native/java/org/jpype/html/AttrParser.java native/java/org/jpype/html/Html.java native/java/org/jpype/html/HtmlGrammar.java native/java/org/jpype/html/HtmlHandler.java native/java/org/jpype/html/HtmlParser.java native/java/org/jpype/html/HtmlTreeHandler.java native/java/org/jpype/html/HtmlWriter.java native/java/org/jpype/html/Parser.java native/java/org/jpype/javadoc/DomUtilities.java native/java/org/jpype/javadoc/Javadoc.java native/java/org/jpype/javadoc/JavadocException.java native/java/org/jpype/javadoc/JavadocExtractor.java native/java/org/jpype/javadoc/JavadocRenderer.java native/java/org/jpype/javadoc/JavadocTransformer.java native/java/org/jpype/manager/ClassDescriptor.java native/java/org/jpype/manager/MethodResolution.java native/java/org/jpype/manager/ModifierCode.java native/java/org/jpype/manager/TypeAudit.java native/java/org/jpype/manager/TypeFactory.java native/java/org/jpype/manager/TypeFactoryNative.java native/java/org/jpype/manager/TypeManager.java native/java/org/jpype/pickle/ByteBufferInputStream.java native/java/org/jpype/pickle/Decoder.java native/java/org/jpype/pickle/Encoder.java native/java/org/jpype/pkg/JPypePackage.java native/java/org/jpype/pkg/JPypePackageManager.java native/java/org/jpype/proxy/JPypeProxy.java native/java/org/jpype/ref/JPypeReference.java native/java/org/jpype/ref/JPypeReferenceNative.java native/java/org/jpype/ref/JPypeReferenceQueue.java native/java/org/jpype/ref/JPypeReferenceSet.java
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Copy file native/java/org/jpype/html/entities.txt build/temp.linux-x86_64-cpython-311/org.jpype/classes/org/jpype/html/entities.txt
  jar cvf build/lib.linux-x86_64-cpython-311/org.jpype.jar -C build/temp.linux-x86_64-cpython-311/org.jpype/classes .
added manifest
adding: org/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/JPypeContext.class(in = 9682) (out= 4714)(deflated 51%)
adding: org/jpype/PyExceptionProxy.class(in = 223) (out= 189)(deflated 15%)
adding: org/jpype/classloader/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/classloader/JPypeClassLoader.class(in = 2649) (out= 1413)(deflated 46%)
adding: org/jpype/classloader/DynamicClassLoader$1.class(in = 1876) (out= 790)(deflated 57%)
adding: org/jpype/classloader/DynamicClassLoader.class(in = 6303) (out= 3073)(deflated 51%)
adding: org/jpype/ref/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/ref/JPypeReferenceQueue.class(in = 3058) (out= 1433)(deflated 53%)
adding: org/jpype/ref/JPypeReferenceSet.class(in = 1628) (out= 927)(deflated 43%)
adding: org/jpype/ref/JPypeReferenceQueue$1.class(in = 185) (out= 143)(deflated 22%)
adding: org/jpype/ref/JPypeReferenceNative.class(in = 267) (out= 204)(deflated 23%)
adding: org/jpype/ref/JPypeReferenceQueue$Worker.class(in = 1560) (out= 748)(deflated 52%)
adding: org/jpype/ref/JPypeReference.class(in = 541) (out= 355)(deflated 34%)
adding: org/jpype/ref/JPypeReferenceSet$Pool.class(in = 690) (out= 405)(deflated 41%)
adding: org/jpype/JPypeContext$1.class(in = 480) (out= 311)(deflated 35%)
adding: org/jpype/JPypeSignal$1.class(in = 576) (out= 360)(deflated 37%)
adding: org/jpype/JPypeKeywords.class(in = 1448) (out= 775)(deflated 46%)
adding: org/jpype/JPypeSignal.class(in = 1711) (out= 820)(deflated 52%)
adding: org/jpype/pkg/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/pkg/JPypePackage.class(in = 4038) (out= 2090)(deflated 48%)
adding: org/jpype/pkg/JPypePackageManager$ModuleDirectory.class(in = 1752) (out= 877)(deflated 49%)
adding: org/jpype/pkg/JPypePackageManager.class(in = 7815) (out= 3609)(deflated 53%)
adding: org/jpype/manager/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/manager/TypeManager$1.class(in = 177) (out= 135)(deflated 23%)
adding: org/jpype/manager/MethodResolution.class(in = 4502) (out= 2186)(deflated 51%)
adding: org/jpype/manager/ClassDescriptor.class(in = 1192) (out= 708)(deflated 40%)
adding: org/jpype/manager/TypeManager$Destroyer.class(in = 1116) (out= 626)(deflated 43%)
adding: org/jpype/manager/TypeManager.class(in = 15280) (out= 7263)(deflated 52%)
adding: org/jpype/manager/TypeFactory.class(in = 644) (out= 332)(deflated 48%)
adding: org/jpype/manager/TypeFactoryNative.class(in = 772) (out= 414)(deflated 46%)
adding: org/jpype/manager/ModifierCode.class(in = 2632) (out= 1425)(deflated 45%)
adding: org/jpype/manager/TypeAudit.class(in = 251) (out= 172)(deflated 31%)
adding: org/jpype/pickle/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/pickle/Decoder.class(in = 683) (out= 406)(deflated 40%)
adding: org/jpype/pickle/Encoder.class(in = 710) (out= 421)(deflated 40%)
adding: org/jpype/pickle/ByteBufferInputStream.class(in = 1179) (out= 649)(deflated 44%)
adding: org/jpype/JPypeUtilities.class(in = 779) (out= 444)(deflated 43%)
adding: org/jpype/proxy/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/proxy/JPypeProxy.class(in = 2535) (out= 1254)(deflated 50%)
adding: org/jpype/html/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/html/AttrGrammar$AttrRule.class(in = 1223) (out= 641)(deflated 47%)
adding: org/jpype/html/AttrGrammar.class(in = 2221) (out= 984)(deflated 55%)
adding: org/jpype/html/Html.class(in = 5891) (out= 3223)(deflated 45%)
adding: org/jpype/html/AttrGrammar$EndQuoteRule.class(in = 1130) (out= 564)(deflated 50%)
adding: org/jpype/html/HtmlGrammar$Comment.class(in = 1744) (out= 805)(deflated 53%)
adding: org/jpype/html/AttrGrammar$QuoteRule.class(in = 1087) (out= 545)(deflated 49%)
adding: org/jpype/html/HtmlGrammar$Quoted.class(in = 1638) (out= 748)(deflated 54%)
adding: org/jpype/html/HtmlGrammar$EndCData.class(in = 1648) (out= 834)(deflated 49%)
adding: org/jpype/html/entities.txt(in = 3616) (out= 1559)(deflated 56%)
adding: org/jpype/html/AttrGrammar$BooleanRule.class(in = 1209) (out= 634)(deflated 47%)
adding: org/jpype/html/Parser$MatchRule.class(in = 1073) (out= 605)(deflated 43%)
adding: org/jpype/html/HtmlGrammar$MergeText.class(in = 996) (out= 525)(deflated 47%)
adding: org/jpype/html/HtmlGrammar$StartElement.class(in = 1624) (out= 784)(deflated 51%)
adding: org/jpype/html/HtmlGrammar$Escaped.class(in = 1048) (out= 556)(deflated 46%)
adding: org/jpype/html/Parser$Token.class(in = 215) (out= 167)(deflated 22%)
adding: org/jpype/html/Parser$Rule.class(in = 351) (out= 197)(deflated 43%)
adding: org/jpype/html/HtmlGrammar$Token.class(in = 2243) (out= 1174)(deflated 47%)
adding: org/jpype/html/HtmlGrammar$Directive.class(in = 910) (out= 476)(deflated 47%)
adding: org/jpype/html/HtmlGrammar$StartQuote.class(in = 1259) (out= 611)(deflated 51%)
adding: org/jpype/html/AttrParser.class(in = 649) (out= 360)(deflated 44%)
adding: org/jpype/html/HtmlParser.class(in = 531) (out= 306)(deflated 42%)
adding: org/jpype/html/HtmlHandler.class(in = 352) (out= 218)(deflated 38%)
adding: org/jpype/html/HtmlTreeHandler.class(in = 5222) (out= 2595)(deflated 50%)
adding: org/jpype/html/AttrGrammar$Token.class(in = 1791) (out= 931)(deflated 48%)
adding: org/jpype/html/HtmlGrammar$BeginElement.class(in = 2130) (out= 962)(deflated 54%)
adding: org/jpype/html/Parser$Grammar.class(in = 254) (out= 165)(deflated 35%)
adding: org/jpype/html/AttrGrammar$IgnoreWSRule.class(in = 819) (out= 458)(deflated 44%)
adding: org/jpype/html/HtmlGrammar$State.class(in = 2080) (out= 905)(deflated 56%)
adding: org/jpype/html/Parser.class(in = 3984) (out= 1900)(deflated 52%)
adding: org/jpype/html/Parser$State.class(in = 349) (out= 198)(deflated 43%)
adding: org/jpype/html/HtmlGrammar$1.class(in = 171) (out= 136)(deflated 20%)
adding: org/jpype/html/HtmlGrammar.class(in = 5251) (out= 2249)(deflated 57%)
adding: org/jpype/html/HtmlGrammar$EndComment.class(in = 1670) (out= 853)(deflated 48%)
adding: org/jpype/html/HtmlGrammar$EndElement.class(in = 1392) (out= 664)(deflated 52%)
adding: org/jpype/html/HtmlGrammar$CompleteElement.class(in = 1709) (out= 853)(deflated 50%)
adding: org/jpype/html/Parser$Entity.class(in = 774) (out= 391)(deflated 49%)
adding: org/jpype/html/HtmlWriter.class(in = 3444) (out= 1779)(deflated 48%)
adding: org/jpype/html/HtmlGrammar$StartComment.class(in = 1554) (out= 723)(deflated 53%)
adding: org/jpype/html/Parser$1.class(in = 161) (out= 132)(deflated 18%)
adding: org/jpype/html/HtmlGrammar$Cleanup.class(in = 2482) (out= 1179)(deflated 52%)
adding: org/jpype/html/HtmlGrammar$EndDirective.class(in = 1404) (out= 677)(deflated 51%)
adding: org/jpype/html/HtmlGrammar$CData.class(in = 1700) (out= 777)(deflated 54%)
adding: org/jpype/html/AttrGrammar$State.class(in = 1830) (out= 776)(deflated 57%)
adding: org/jpype/javadoc/(in = 0) (out= 0)(stored 0%)
adding: org/jpype/javadoc/JavadocTransformer.class(in = 9712) (out= 4497)(deflated 53%)
adding: org/jpype/javadoc/JavadocExtractor.class(in = 6448) (out= 3058)(deflated 52%)
adding: org/jpype/javadoc/JavadocException.class(in = 277) (out= 198)(deflated 28%)
adding: org/jpype/javadoc/Javadoc.class(in = 626) (out= 346)(deflated 44%)
adding: org/jpype/javadoc/DomUtilities.class(in = 3165) (out= 1450)(deflated 54%)
adding: org/jpype/javadoc/JavadocTransformer$Workspace.class(in = 916) (out= 430)(deflated 53%)
adding: org/jpype/javadoc/JavadocRenderer.class(in = 7226) (out= 3569)(deflated 50%)
Call build ext
building '_jpype' extension
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_array.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_array.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_arrayclass.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_arrayclass.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_booleantype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_booleantype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_boxedtype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_boxedtype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_buffer.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_buffer.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_buffertype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_buffertype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_bytetype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_bytetype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_chartype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_chartype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_class.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_class.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_classhints.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_classhints.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_classloader.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_classloader.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_classtype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_classtype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_context.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_context.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_convert.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_convert.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_doubletype.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_doubletype.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_encoding.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_encoding.o -g0 -std=c++11 -O2
ccache gcc -Wsign-compare -DNDEBUG -fwrapv -Wall -fPIC -Inative/common/include -Inative/python/include -Inative/embedded/include -Inative/jni_include -I/opt/_internal/cpython-3.11.0rc1/include/python3.11 -c native/common/jp_exception.cpp -o build/temp.linux-x86_64-cpython-311/native/common/jp_exception.o -g0 -std=c++11 -O2
native/common/jp_exception.cpp: In function ‘PyTracebackObject* tb_create(PyTracebackObject*, PyObject*, const char*, const char*, int)’:
native/common/jp_exception.cpp:524:9: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘frame’; did you mean ‘cframe’?
  524 |   state.frame = last_traceback->tb_frame;
      |         ^~~~~
      |         cframe
native/common/jp_exception.cpp:526:9: error: ‘PyThreadState’ {aka ‘struct _ts’} has no member named ‘frame’; did you mean ‘cframe’?
  526 |   state.frame = NULL;
      |         ^~~~~
      |         cframe
native/common/jp_exception.cpp:552:29: error: invalid use of incomplete type ‘PyFrameObject’ {aka ‘struct _frame’}
  552 |  traceback->tb_lasti = frame->f_lasti;
      |                             ^~
In file included from /opt/_internal/cpython-3.11.0rc1/include/python3.11/Python.h:42,
                 from native/common/jp_exception.cpp:16:
/opt/_internal/cpython-3.11.0rc1/include/python3.11/pytypedefs.h:22:16: note: forward declaration of ‘PyFrameObject’ {aka ‘struct _frame’}
   22 | typedef struct _frame PyFrameObject;
      |                ^~~~~~
error: command '/usr/bin/ccache' failed with exit code 1
The command '/bin/sh -c cd /opt/src  && CC="ccache gcc" python3.11 setup.py build_ext --inplace ;' returned a non-zero code: 1
```

Sorry about that.  Yes it appears that I checked in to Thrameos/master rather than the branch I wanted.  Current I show one fail...

```========================================================================================================= FAILURES =========================================================================================================
__________________________________________________________________________________________ JCallerSensitiveCase.testStackWalker2 ___________________________________________________________________________________________

>   ???
E   Exception: Java Exception

Caller.java:78: Exception

The above exception was the direct cause of the following exception:

self = <test_caller_sensitive.JCallerSensitiveCase testMethod=testStackWalker2>

    def testStackWalker2(self):
>       self.assertEqual(self.obj.callStackWalker2(), jpype.JClass(
            jpype.java.lang.Class.forName("org.jpype.JPypeContext")).class_)
E       java.lang.java.lang.IllegalCallerException: java.lang.IllegalCallerException: no caller frame

test/jpypetest/test_caller_sensitive.py:111: java.lang.IllegalCallerException```

--- a/doc/CHANGELOG.rst
+++ b/doc/CHANGELOG.rst
@@ -8,6 +8,8 @@ Latest Changes:
   
   - Fixed issue with startJVM changing locale settings.
 
+  - Changes to support Python 3.11
+
 - **1.4.0 - 2022-05-14**
 
   - Support for all different buffer type conversions.

--- a/jpype/_jvmfinder.py
+++ b/jpype/_jvmfinder.py
@@ -320,10 +320,10 @@ class DarwinJVMFinder(LinuxJVMFinder):
         """
         import platform
         import subprocess
-        from distutils.version import StrictVersion
+        from packaging.version import Version
 
-        current = StrictVersion(platform.mac_ver()[0][:4])
-        if current >= StrictVersion('10.6') and current < StrictVersion('10.9'):
+        current = Version(platform.mac_ver()[0][:4])
+        if current >= Version('10.6') and current < Version('10.9'):
             return subprocess.check_output(
                 ['/usr/libexec/java_home']).strip()
 

--- a/native/common/jp_context.cpp
+++ b/native/common/jp_context.cpp
@@ -282,14 +282,10 @@ void JPContext::initializeResources(JNIEnv* env, bool interrupt)
 
 	if (!m_Embedded)
 	{
-		JPPyObject import = JPPyObject::call(PyImport_AddModule("importlib.util"));
+		JPPyObject import = JPPyObject::use(PyImport_AddModule("importlib.util"));
 		JPPyObject jpype = JPPyObject::call(PyObject_CallMethod(import.get(), "find_spec", "s", "_jpype"));
 		JPPyObject origin = JPPyObject::call(PyObject_GetAttrString(jpype.get(), "origin"));
 		val[2].l = frame.fromStringUTF8(JPPyString::asStringUTF8(origin.get()));
-		import.incref();  // The documentation specifies that PyImport_AddModule must return a 
-	                  // new reference, but that is not happening in Python 3.10
-	                  // so we are triggering a gc assertion failure.  To prevent
-	                  // the failure manually up the reference counter here.
 	}
 	m_JavaContext = JPObjectRef(frame, frame.CallStaticObjectMethodA(contextClass, startMethod, val));
 

--- a/native/common/jp_exception.cpp
+++ b/native/common/jp_exception.cpp
@@ -508,12 +508,6 @@ PyObject *tb_create(
 	PyFrameObject *pframe = PyFrame_New(state, (PyCodeObject*) code.get(), dict, NULL);
 	JPPyObject frame = JPPyObject::accept((PyObject*)pframe);
 
-	// Swap the back pointer
-	//PyObject* old = frame->f_back;
-	//frame->f_back = last_traceback->tb_frame;
-	//Py_XDECREF(old);
-	//Py_XINCREF(last_traceback->tb_frame);
-	
 	// If we don't get the frame object there is no point
 	if (frame.get() == NULL)
 		return NULL;

--- a/native/common/jp_exception.cpp
+++ b/native/common/jp_exception.cpp
@@ -519,7 +519,11 @@ PyObject *tb_create(
 		return NULL;
 
 	// Create a traceback
+#if PY_VERSION_HEX<0x03110000
+	JPPyObject lasti = JPPyObject::claim(PyLong_FromLong(pframe->f_lasti));
+#else
 	JPPyObject lasti = JPPyObject::claim(PyLong_FromLong(PyFrame_GetLasti(pframe)));
+#endif
 	JPPyObject linenuma = JPPyObject::claim(PyLong_FromLong(linenum));
 	JPPyObject tuple = JPPyObject::call(PyTuple_Pack(4, Py_None, frame.get(), lasti.get(), linenuma.get()));
 	JPPyObject traceback = JPPyObject::accept(PyObject_Call((PyObject*) &PyTraceBack_Type, tuple.get(), NULL));

--- a/native/common/jp_exception.cpp
+++ b/native/common/jp_exception.cpp
@@ -489,18 +489,18 @@ void JPypeException::toJava(JPContext *context)
 	JP_TRACE_OUT; // GCOVR_EXCL_LINE
 }
 
-PyTracebackObject *tb_create(
-		PyTracebackObject *last_traceback,
+PyObject *tb_create(
+		PyObject *last_traceback,
 		PyObject *dict,
 		const char* filename,
 		const char* funcname,
 		int linenum)
 {
 	// Create a code for this frame. (ref count is 1)
-	PyCodeObject *code = PyCode_NewEmpty(filename, funcname, linenum);
+	JPPyObject code = JPPyObject::accept((PyObject*)PyCode_NewEmpty(filename, funcname, linenum));
 
 	// If we don't get the code object there is no point
-	if (code == NULL)
+	if (code.get() == NULL)
 		return NULL;
 
 	// This is a bit of a kludge.  Python lacks a way to directly create
@@ -520,48 +520,40 @@ PyTracebackObject *tb_create(
 	// portable, but we have to create a big (uninitialized object) each time we
 	// want to pass in the previous frame.
 	PyThreadState state;
+	JPPyObject prev;
 	if (last_traceback != NULL)
-		state.frame = last_traceback->tb_frame;
+	{
+       		prev = JPPyObject::call(PyObject_GetAttrString(last_traceback, "tb_frame"));
+		state.frame = (PyFrameObject*) prev.get();
+	}
 	else
 		state.frame = NULL;
 
 	// Create a frame for the traceback.
-	PyFrameObject *frame = PyFrame_New(&state, code, dict, NULL);
-	
-	// frame just borrows the reference rather than claiming it
-	// so we need to get rid of the extra reference here.
-	Py_DECREF(code);
+	JPPyObject frame = JPPyObject::call((PyObject*) PyFrame_New(&state, (PyCodeObject*) code.get(), dict, NULL));
 	
 	// If we don't get the frame object there is no point
-	if (frame == NULL)
+	if (frame.get() == NULL)
 		return NULL;
 
 	// Create a traceback
-	PyTracebackObject *traceback = (PyTracebackObject*) 
-		        PyObject_GC_New(PyTracebackObject, &PyTraceBack_Type);
+	JPPyObject lasti = JPPyObject::claim(PyLong_FromLong(((PyFrameObject*)frame.get())->f_lasti));
+	JPPyObject linenuma = JPPyObject::claim(PyLong_FromLong(linenum));
+	JPPyObject tuple = JPPyObject::call(PyTuple_Pack(4, Py_None, frame.get(), lasti.get(), linenuma.get()));
+	JPPyObject traceback = JPPyObject::accept(PyObject_Call((PyObject*) &PyTraceBack_Type, tuple.get(), NULL));
 
 	// We could fail in process
-	if (traceback == NULL)
+	if (traceback.get() == NULL)
 	{
-		Py_DECREF(frame);
 		return NULL;
 	}
 
-	// Set the fields
-	traceback->tb_frame = frame; // Steal the reference from frame
-	traceback->tb_lasti = frame->f_lasti;
-	traceback->tb_lineno = linenum;
-	Py_XINCREF(last_traceback);
-	traceback->tb_next = last_traceback;
-
-	// Allow GC on the object
-	PyObject_GC_Track(traceback);
-	return traceback;
+	return traceback.keep();
 }
 
 PyObject* PyTrace_FromJPStackTrace(JPStackTrace& trace)
 {
-	PyTracebackObject *last_traceback = NULL;
+	PyObject *last_traceback = NULL;
 	PyObject *dict = PyModule_GetDict(PyJPModule);
 	for (JPStackTrace::iterator iter = trace.begin(); iter != trace.end(); ++iter)
 	{
@@ -575,7 +567,7 @@ PyObject* PyTrace_FromJPStackTrace(JPStackTrace& trace)
 
 JPPyObject PyTrace_FromJavaException(JPJavaFrame& frame, jthrowable th, jthrowable prev)
 {
-	PyTracebackObject *last_traceback = NULL;
+	PyObject *last_traceback = NULL;
 	JPContext *context = frame.getContext();
 	jvalue args[2];
 	args[0].l = th;

--- a/native/common/jp_exception.cpp
+++ b/native/common/jp_exception.cpp
@@ -503,30 +503,16 @@ PyTracebackObject *tb_create(
 	if (code == NULL)
 		return NULL;
 
-	// This is a bit of a kludge.  Python lacks a way to directly create
-	// a frame from a code object except when creating from the threadstate.
-	//
-	// In reviewing Python implementation, I find that the only element accessed
-	// in the thread state was the previous frame to link to.  Because frame 
-	// objects change a lot between different Python versions, trying to 
-	// replicate the actions of setting up a frame is difficult to keep portable.
-	//
-	// Python 3.10 introduces the additional requirement that the global
-	// dictionary supplied must have a __builtins__.  We can do this once
-	// when create the module.
-	//
-	// If instead we create a thread state and point the field it needs to the
-	// previous frame we create the frames using the defined API.  Much more 
-	// portable, but we have to create a big (uninitialized object) each time we
-	// want to pass in the previous frame.
-	PyThreadState state;
-	if (last_traceback != NULL)
-		state.frame = last_traceback->tb_frame;
-	else
-		state.frame = NULL;
 
 	// Create a frame for the traceback.
-	PyFrameObject *frame = PyFrame_New(&state, code, dict, NULL);
+	PyThreadState *state = PyThreadState_GET();
+	PyFrameObject *frame = PyFrame_New(state, code, dict, NULL);
+
+	// Swap the back pointer
+	//PyObject* old = frame->f_back;
+	//frame->f_back = last_traceback->tb_frame;
+	//Py_XDECREF(old);
+	//Py_XINCREF(last_traceback->tb_frame);
 	
 	// frame just borrows the reference rather than claiming it
 	// so we need to get rid of the extra reference here.
@@ -549,7 +535,7 @@ PyTracebackObject *tb_create(
 
 	// Set the fields
 	traceback->tb_frame = frame; // Steal the reference from frame
-	traceback->tb_lasti = frame->f_lasti;
+	traceback->tb_lasti = PyFrame_GetLasti(frame);
 	traceback->tb_lineno = linenum;
 	Py_XINCREF(last_traceback);
 	traceback->tb_next = last_traceback;

--- a/native/python/pyjp_value.cpp
+++ b/native/python/pyjp_value.cpp
@@ -43,8 +43,34 @@ PyObject* PyJPValue_alloc(PyTypeObject* type, Py_ssize_t nitems )
 	JP_PY_TRY("PyJPValue_alloc");
 	// Modification from Python to add size elements
 	const size_t size = _PyObject_VAR_SIZE(type, nitems + 1) + sizeof (JPValue);
-	PyObject *obj = (PyType_IS_GC(type)) ? _PyObject_GC_Malloc(size)
-			: (PyObject *) PyObject_MALLOC(size);
+	PyObject *obj = NULL;
+	if (PyType_IS_GC(type))
+	{
+		// Horrible kludge because python lacks an API for allocating a GC type with extra memory
+		// The privor method _PyObject_GC_Alloc is no longer visible, so we are forced to allocate
+		// a different type with the extra memory and then hot swap the type to the real one.
+		PyTypeObject type2;
+		type2.tp_basicsize = size;
+		type2.tp_itemsize = 0;
+		type2.tp_name = NULL;
+		type2.tp_flags = type->tp_flags;
+		type2.tp_traverse = type->tp_traverse;
+
+		// Allocate the fake type
+		obj = PyObject_GC_New(PyObject, &type2);
+
+		// Hot swap to the required type
+		obj->ob_type = type;
+		if (PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE))
+		{
+			Py_INCREF(type);
+		}
+		// Note the object will be inited twice which should not leak. (fingers crossed)
+	}
+	else
+	{
+		obj = (PyObject*) PyObject_MALLOC(size);
+	}
 	if (obj == NULL)
 		return PyErr_NoMemory(); // GCOVR_EXCL_LINE
 	memset(obj, 0, size);

--- a/native/python/pyjp_value.cpp
+++ b/native/python/pyjp_value.cpp
@@ -23,8 +23,6 @@ extern "C"
 {
 #endif
 
-extern PyObject * _PyObject_GC_Malloc(size_t basicsize);
-
 /**
  * Allocate a new Python object with a slot for Java.
  *
@@ -49,7 +47,7 @@ PyObject* PyJPValue_alloc(PyTypeObject* type, Py_ssize_t nitems )
 	if (PyType_IS_GC(type))
 	{
 		// Horrible kludge because python lacks an API for allocating a GC type with extra memory
-		// The privor method _PyObject_GC_Alloc is no longer visible, so we are forced to allocate
+		// The private method _PyObject_GC_Alloc is no longer visible, so we are forced to allocate
 		// a different type with the extra memory and then hot swap the type to the real one.
 		PyTypeObject type2;
 		type2.tp_basicsize = size;

--- a/native/python/pyjp_value.cpp
+++ b/native/python/pyjp_value.cpp
@@ -61,12 +61,6 @@ PyObject* PyJPValue_alloc(PyTypeObject* type, Py_ssize_t nitems )
 		// Allocate the fake type
 		obj = PyObject_GC_New(PyObject, &type2);
 
-		// Hot swap to the required type
-		obj->ob_type = type;
-		if (PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE))
-		{
-			Py_INCREF(type);
-		}
 		// Note the object will be inited twice which should not leak. (fingers crossed)
 	}
 	else

--- a/native/python/pyjp_value.cpp
+++ b/native/python/pyjp_value.cpp
@@ -23,6 +23,8 @@ extern "C"
 {
 #endif
 
+extern PyObject * _PyObject_GC_Malloc(size_t basicsize);
+
 /**
  * Allocate a new Python object with a slot for Java.
  *

--- a/setup.py
+++ b/setup.py
@@ -78,7 +78,8 @@ setup(
     packages=['jpype', 'jpype._pyinstaller'],
     package_dir={'jpype': 'jpype', },
     package_data={'jpype': ['*.pyi']},
-    install_requires=['typing_extensions ; python_version< "3.8"'],
+    install_requires=['typing_extensions ; python_version< "3.8"',
+        'packaging ; python_version< "3.10"'],
     tests_require=['pytest'],
     extras_require={
         'tests': [

--- a/setupext/test_java.py
+++ b/setupext/test_java.py
@@ -53,7 +53,7 @@ def compileJava():
     srcs = glob.glob('test/harness/jpype/**/*.java', recursive=True)
     srcs.extend(glob.glob('test/harness/org/**/*.java', recursive=True))
     exports = ""
-    if version > 7:
+    if version == 8:
         srcs.extend(glob.glob('test/harness/java8/**/*.java', recursive=True))
     if version > 8:
         srcs.extend(glob.glob('test/harness/java9/**/*.java', recursive=True))

